caput/mpiarray.py

```diff
@@ -1602,97 +1610,226 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
-            kwargs["out"] = tuple(out_args)
-        else:
-            outputs = (None,) * ufunc.nout
+        # Where arguments are not supported, except where the value is a simple `True`,
+        # which is used as a default
+        if "where" in kwargs and not (
+            isinstance(kwargs["where"], bool) and kwargs["where"]
+        ):
+            raise UnsupportedOperation(
+                "MPIArray ufunc calls do not support 'where' arguments, but "
+                f"where={kwargs['where']} was found. Try using numpy arrays directly "
+                "if you really need to use a 'where' argument."
+            )
 
-        results = super().__array_ufunc__(ufunc, method, *args, **kwargs)
+        comm = _get_common_comm(inputs)
 
-        # that ufunc was not implemented for ndarrays
-        if results is NotImplemented:
-            return NotImplemented
+        new_dist_axis, global_length, offset = _validation_fn(ufunc, *inputs, **kwargs)
 
-        # operation was performed in-place, so we can just return
-        if method == "at":
-            return
+        # Check that all out arguments are valid MPIArrays, and then convert into numpy
+        # arrays
+        if "out" in kwargs:
+            out_args = kwargs.get("out", None)
+            kwargs["out"] = _mpi_to_ndarray(out_args, only_mpiarray=True)
 
-        if ufunc.nout == 1:
-            results = (results,)
+            # Check the distribution of the output arguments makes sense
+            for ii, array in enumerate(out_args):
+                try:
+                    _check_dist_axis(array, new_dist_axis, global_length, offset, comm)
+                except Exception as e:
+                    raise ValueError(
+                        f"Output argument at position {ii} does not match expectation."
+                    ) from e
+        else:
+            out_args = [None] * ufunc.nout
 
-        if "reduce" in method and (
-            results[0].shape or getattr(outputs[0], "shape", None)
-        ):
-            # reduction methods eliminate axes, so the distributed axis
-            # might need to be recalculated
-            # except when the user explicitly specifies keepdims
-            if not kwargs.get("keepdims", False) and (kwargs["axis"] < dist_axis):
-                dist_axis -= 1
+        results = _ensure_list(
+            super().__array_ufunc__(ufunc, method, *_mpi_to_ndarray(inputs), **kwargs)
+        )
 
+        # Convert any outputs back into valid MPIArrays if required
         ret = []
+        for res, out in zip(results, out_args):
+            # If out is None a new *numpy* array was initialised for the results. We
+            # need to turn this into an MPIArray using the inferrred distribution
+            if out is None:
+                # Result is a scalar, so we want to return this as a 1D distributed
+                # vector
+                if not res.shape:
+                    res = res.reshape(1)
+
```

Comment:
Just confirming that all values returned from super().__array_ufunc__ are numpy objects, otherwise this could fail (if something doesn't have a .reshape routine)

Reply:
Good point. However, I think as you suggest it is guaranteed that even scalar values are numpy objects and have both `.shape` and `.reshape()`.